modules/common/tls/tls.go

```diff
@@ -20,145 +20,403 @@ package tls
+type GenericService struct {
 	// +kubebuilder:validation:Optional
-	DisableNonTLSListeners bool `json:"disableNonTLSListeners,omitempty"`
+	// SecretName - holding the cert, key for the service
+	SecretName *string `json:"secretName,omitempty"`
 }
 
 // Ca contains CA-specific settings, which could be used both by services (to define their own CA certificates)
 // and by clients (to verify the server's certificate)
 type Ca struct {
-	// +kubebuilder:validation:Optional
-	CaSecretName string `json:"caSecretName,omitempty"`
+	// CaBundleSecretName - holding the CA certs in a pre-created bundle file
+	CaBundleSecretName string `json:"caBundleSecretName,omitempty"`
 }
 
-// TLS - a generic type, which encapsulates both the service and CA configurations
-type TLS struct {
-	Service *Service `json:"service"`
-	Ca      *Ca      `json:"ca"`
+// Service contains server-specific TLS secret
+// +kubebuilder:object:generate:=false
+type Service struct {
+	// SecretName - holding the cert, key for the service
+	SecretName string `json:"secretName"`
+
+	// CertMount - dst location to mount the service tls.crt cert. Can be used to override the default location which is /etc/tls/certs/<service id>.crt
+	CertMount *string `json:"certMount,omitempty"`
+
+	// KeyMount - dst location to mount the service tls.key  key. Can be used to override the default location which is /etc/tls/private/<service id>.key
+	KeyMount *string `json:"keyMount,omitempty"`
+
+	// CaMount - dst location to mount the CA cert ca.crt to. Can be used if the service CA cert should be mounted specifically, e.g. to be set in a service config for validation, instead of the env wide bundle.
+	CaMount *string `json:"caMount,omitempty"`
+}
+
+// Enabled - returns true if tls is configured for the public and internal
+func (a *APIService) Enabled(endpt service.Endpoint) bool {
+	switch endpt {
+	case service.EndpointPublic:
+		return (a.Disabled == nil || (a.Disabled != nil && !*a.Disabled)) && a.Public.SecretName != nil
+	case service.EndpointInternal:
+		return (a.Disabled == nil || (a.Disabled != nil && !*a.Disabled)) && a.Internal.SecretName != nil
+	}
+
+	return false
 }
 
-// NewTLS - initialize and return a TLS struct
-func NewTLS(ctx context.Context, h *helper.Helper, namespace string, service *Service, ca *Ca) (*TLS, error) {
+// ValidateCertSecrets - validates the content of the cert secrets to make sure "tls-ca-bundle.pem" key exist
+func (a *APIService) ValidateCertSecrets(
+	ctx context.Context,
+	h *helper.Helper,
+	namespace string,
+) (string, ctrl.Result, error) {
+	var svc GenericService
+	certHashes := map[string]env.Setter{}
+	for _, endpt := range []service.Endpoint{service.EndpointInternal, service.EndpointPublic} {
+		switch endpt {
+		case service.EndpointPublic:
+			if !a.Enabled(service.EndpointPublic) {
+				continue
+			}
+
+			svc = a.Public
+
+		case service.EndpointInternal:
+			if !a.Enabled(service.EndpointInternal) {
+				continue
+			}
+
+			svc = a.Public
```

Comment:
Is this right?  Looks the same as the public endpoint.

Reply:
no this is wrong, good catch